pkg/bandwidth/bandwidth.go

```diff
@@ -0,0 +1,159 @@
+	"github.com/cilium/cilium/pkg/logging"
+	"github.com/cilium/cilium/pkg/logging/logfields"
+	"github.com/cilium/cilium/pkg/maps/bwmap"
+	"github.com/cilium/cilium/pkg/option"
+	"github.com/cilium/cilium/pkg/sysctl"
+	"github.com/sirupsen/logrus"
+	"github.com/vishvananda/netlink"
+
+	"k8s.io/apimachinery/pkg/api/resource"
+)
+
+const (
+	subsystem = "bandwidth-manager"
+
+	// EgressBandwidth is the K8s Pod annotation.
+	EgressBandwidth = "kubernetes.io/egress-bandwidth"
+)
+
+var log = logging.DefaultLogger.WithField(logfields.LogSubsys, subsystem)
+
+func GetBytesPerSec(bandwidth string) (uint64, error) {
+	res, err := resource.ParseQuantity(bandwidth)
+	if err != nil {
+		return 0, err
+	}
+	return uint64(res.Value() / 8), err
+}
+
+func ProbeBandwidthManager() {
+	if option.Config.DryMode || !option.Config.EnableBandwidthManager {
+		return
+	}
+
+	if _, err := sysctl.Read("net.core.default_qdisc"); err != nil {
+		log.Warn("BPF bandwidth manager could not read procfs. Disabling the feature.")
+		option.Config.EnableBandwidthManager = false
+		return
+	}
+
+	kernelGood := false
+	if h := probes.NewProbeManager().GetHelpers("sched_cls"); h != nil {
+		// We at least need 5.0 kernel for native TCP EDT integration.
+		// Below helper is available for 5.0 kernels and onwards.
+		if _, ok := h["bpf_sk_release"]; ok {
+			kernelGood = true
+		}
+	}
+	if !kernelGood {
+		log.Warn("BPF bandwidth manager needs kernel 5.0 or newer. Disabling the feature.")
+		option.Config.EnableBandwidthManager = false
+		return
+	}
+}
+
+func InitBandwidthManager() {
+	if option.Config.DryMode || !option.Config.EnableBandwidthManager {
+		return
+	}
+
+	if len(option.Config.Devices) == 0 {
+		log.Warn("BPF bandwidth manager could not detect host devices. Disabling the feature.")
+		option.Config.EnableBandwidthManager = false
+		return
+	}
+
+	log.Info("Setting up BPF bandwidth manager")
+
+	if _, err := bwmap.ThrottleMap.OpenOrCreate(); err != nil {
+		log.WithError(err).Fatal("Failed to access ThrottleMap")
+	}
+
+	type setting struct {
+		name string
+		val  string
+	}
+	baseSettings := []setting{
+		{"net.core.netdev_max_backlog", "1000"},
+		{"net.core.somaxconn", "4096"},
+		{"net.core.default_qdisc", "fq"},
+		{"net.ipv4.tcp_congestion_control", "bbr"},
+		{"net.ipv4.tcp_max_syn_backlog", "4096"},
+	}
+	for _, s := range baseSettings {
+		log.Infof("Setting sysctl %s=%s", s.name, s.val)
+		if err := sysctl.Write(s.name, s.val); err != nil {
+			log.WithError(err).WithFields(logrus.Fields{
+				logfields.SysParamName:  s.name,
+				logfields.SysParamValue: s.val,
+			}).Fatal("Failed to set sysctl needed by BPF bandwidth manager.")
+		}
+	}
+
+	for _, device := range option.Config.Devices {
+		link, err := netlink.LinkByName(device)
+		if err != nil {
+			log.WithError(err).WithField("device", device).Warn("Link does not exist")
+			continue
+		}
+		// We strictly want to avoid a down/up cycle on the device at
```

Comment:
Q: this means that a previously attached bpf_host won't get temporarily detached?

Reply:
No, this is the actual tc 'egress' qdisc which won't affect sch_clsact egress.